llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -1405,6 +1405,14 @@ static Instruction *foldBoxMultiply(BinaryOperator &I) {
   // ResLo = (CrossSum << HalfBits) + (YLo * XLo)
   Value *XLo, *YLo;
   Value *CrossSum;
+  
+  // Checking the operands of I is used in no more than one place,
+  // which can not be deleted, cause a mul instruction has far more weight than
+  // add and shl instruction in IR, thus this method cannot achieve the goal of
+  // simplifying instructions, just return null.
+  if ((!I.getOperand(0)->hasOneUser() || !I.getOperand(1)->hasOneUser()))
+    return nullptr;
+
   if (!match(&I, m_c_Add(m_Shl(m_Value(CrossSum), m_SpecificInt(HalfBits)),
                          m_Mul(m_Value(YLo), m_Value(XLo)))))
```

Comment:
```suggestion
  if (!match(&I, m_c_Add(m_Shl(m_Value(CrossSum), m_SpecificInt(HalfBits)),
                         m_OneUse(m_Mul(m_Value(YLo), m_Value(XLo))))))
```
I think applying the one-use constraint on mul inst is enough.
See also https://reviews.llvm.org/D136015#3865898.

Reply:
> I think applying the one-use constraint on mul inst is enough. See also https://reviews.llvm.org/D136015#3865898.

I think the purpose of box multiply fold optimization is to minimize the number of mul instruction. If the shl instruction does not guarantee one-use, the number of mul instruction will not decrease in the end. The meaning of the combine does not achieve the expected effect.

Reply:
If (only) the mul is one-use, the number of multiplies will stay the same, but the new multiply is more amenable to further optimization.

Reply:
I am not sure the scenario of the further optimization，can you give some examples?

Reply:
here's an example https://gcc.godbolt.org/z/fz9GT3r7Y, we hope to combine the `high`（todo） and `low` part

Reply:
> I think applying the one-use constraint on mul inst is enough. See also https://reviews.llvm.org/D136015#3865898.
I agree，add inst + mul inst will be transformed to mul inst，it will bring profit.